engine/table/src/main/java/io/deephaven/engine/table/impl/MultiJoinTableImpl.java

```diff
@@ -0,0 +1,671 @@
+                .toArray(ColumnSource[]::new);
+    }
+
+    @NotNull
+
+    private static ColumnSource<?>[] getOriginalKeyColumns(
+            @NotNull final MultiJoinInput joinInput) {
+        return Arrays.stream(joinInput.columnsToMatch())
+                .map(jm -> joinInput.inputTable().getColumnSource(jm.right().name()))
+                .toArray(ColumnSource<?>[]::new);
+    }
+
+    private static class MultiJoinListenerRecorder extends ListenerRecorder {
+        private final ColumnSource<?>[] keyColumns;
+        private final ModifiedColumnSet sourceKeyModifiedColumnSet;
+        private final ModifiedColumnSet.Transformer transformer;
+        private final int tableNumber;
+
+        public MultiJoinListenerRecorder(@NotNull final String description,
+                @NotNull final QueryTable parent,
+                @NotNull final QueryTable dependent,
+                final ColumnSource<?>[] keyColumns,
+                final ModifiedColumnSet sourceKeyModifiedColumnSet,
+                @NotNull final ModifiedColumnSet.Transformer transformer,
+                final int tableNumber) {
+            super(description, parent, dependent);
+            this.keyColumns = keyColumns;
+            this.sourceKeyModifiedColumnSet = sourceKeyModifiedColumnSet;
+            this.transformer = transformer;
+            this.tableNumber = tableNumber;
+        }
+
+        @Override
+        public Table getParent() {
+            return super.getParent();
+        }
+    }
+
+    private static class MultiJoinMergedListener extends MergedListener {
+        private final IncrementalMultiJoinStateManagerTypedBase stateManager;
+        private final List<MultiJoinListenerRecorder> recorders;
+        private final ModifiedColumnSet[] modifiedColumnSets;
+        private final MultiJoinModifiedSlotTracker slotTracker = new MultiJoinModifiedSlotTracker();
+
+        protected MultiJoinMergedListener(@NotNull final IncrementalMultiJoinStateManagerTypedBase stateManager,
+                @NotNull final List<MultiJoinListenerRecorder> recorders,
+                @NotNull final Collection<NotificationQueue.Dependency> dependencies,
+                @NotNull final String listenerDescription,
+                @NotNull final QueryTable result,
+                @NotNull final ModifiedColumnSet[] modifiedColumnSets) {
+            super(recorders, dependencies, listenerDescription, result);
+            this.stateManager = stateManager;
+            this.recorders = recorders;
+            this.modifiedColumnSets = modifiedColumnSets;
+        }
+
+        @Override
+        protected void process() {
+            slotTracker.clear();
+            slotTracker.ensureTableCapacity(stateManager.getTableCount());
+
+            final long originalSize = stateManager.getResultSize();
+
+            for (MultiJoinListenerRecorder recorder : recorders) {
+                if (recorder.recordedVariablesAreValid()) {
```

Comment:
I think we can ignore recorded updates if there are only modifies to columns that are not key columns and not join additions. That might be worthwhile. That does, however, create a possibility that we might have to *not* deliver a downstream notification in some cases, because it will be empty.